storage: improve replica rebalancing

cuongdo opened the pull request:
Previously, the replica allocator was choosing rebalancing targets based
on a combination of used store capacity and range counts. In practice,
for small clusters, this didn't rebalance any replicas.

This change makes the allocator spread replicas as evenly as possible
across the stores that have sufficient available capacity.

I've also added intentional delays between rebalances, because it doesn't
make much sense to do more than one rebalance per store gossip period
(currently set to 1 minute). Any more than that risks unnecessary
replica rebalancing caused by outdated gossiped StoreDescriptors.

Resolves #5692.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6133)

<!-- Reviewable:end -->

cuongdo commented:
Here's a graph that shows the backoff interval for replica rebalances vs. standard deviation of the range count:

![screenshot 2016-04-18 16 33 59](https://cloud.githubusercontent.com/assets/1564582/14620031/8fb50a40-0588-11e6-84e3-bd3afd67a0e8.png)

BramGruneir commented:
I'll try to take a look tonight or tomorrow morning.

petermattis commented on storage/allocator.go:
```diff
@@ -35,21 +37,16 @@ const (
+	// minSecsBetweenRebalances is the minimum number of seconds that must elapse
+	// between replica rebalances. This should be slightly larger than
+	// server.gossipStoresInterval.
+	minRebalanceInterval = 65 * time.Second
```
I'm not sure I believe this will do the right thing. Imagine a 3-node cluster where each node has 100 replicas. Then a 4th node is added. We want to move 25 replicas from each existing node (75 replicas total) "quickly". With this setting, "quickly" will take ~30 min if the range leadership is evenly spread amongst the existing nodes and ~80 min if it is unbalanced. Another perspective: 75 replicas at 64 MB per replicas is ~700 MB of data. We should be able to rebalance that much data in a handful of minutes.

cuongdo replied:
I talked to Pete in person, but it's useful to document some of that here (and the rest in an RFC).

There's a tension here between avoiding flip-flopping of replicas between nodes and speeding things up. Because we only gossip `StoreDescriptor`s once a minute, there's always a risk that the allocator is using outdated data, which causes something like a feedback loop.

Let's take your example:

1. Stores 1, 2, 3, each have 100 replicas
2. You add store 4, which is empty.
3. Within the first ~1 minute after adding store 4, the allocators for stores 1-3 will think that store 4 is empty based on the gossiped `StoreDescriptor`.
4. Store 4 gets too many replicas (again, because it appears to have 0 replicas based on outdated gossip info)
5. Store 4 needs to rebalance replicas back to stores 1-3 to even things out.
6. Lots of flip-flopping of replicas happens.
7. In my own experiments, it takes several *thousand* range movements for things to stabilize in this scenario. Using the slower but surer heuristic implemented here, things stabilize in < 100 range events.

tschottdorf commented:
Haven't closely looked at the change, but one question.

> This change makes the allocator spread replicas as evenly as possible
across the stores that have sufficient available capacity.

In large clusters, could that lead to flip-flopping rebalancing or will such a cluster still be relatively stable?

cuongdo commented:
One of my goals was to avoid flip-flopping.

Let's take a trivial case:

Store 1: 3 replicas
Store 2: 3 replicas
Store 3: 4 replicas
Store 4: 3 replicas

(Ignore for a second that with 3-way replication, this scenario shouldn't
exist for long.)

Store 3 will not rebalance any replicas to any other stores, because the
chosen heuristic ensures (as much as possible with potentially outdated
gossiped `StoreDescriptor`s) that any rebalance must *reduce* overall
variance in range counts. We won't rebalance from stores 1, 2, and 4
either, because moving replicas away from those stores will increase
variance.

On Tue, Apr 19, 2016 at 9:36 AM Tobias Schottdorf <notifications@github.com>
wrote:

> Haven't closely looked at the change, but one question.
>
> This change makes the allocator spread replicas as evenly as possible
> across the stores that have sufficient available capacity.
>
> In large clusters, could that lead to flip-flopping rebalancing or will
> such a cluster still be relatively stable?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6133#issuecomment-211925428>
>

BramGruneir commented:
I like this change.  A great first step in updating the rebalancing.  
Is there any way to add a bit more testing?  For trying to rebalance too quickly say.

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/allocator.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFfC_ys5EFUmT-ZUVlI-r1-49:-KFg79Y-CJwM60sgzwzR:1475051600) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L49)):*
Why not peg the value to the gossipStoresInterval by default if you require it to be larger?  Say + 1 or 5 second.  Or even using the value directly?
For testing purposes, it's probably worth piping this value in via an environment variable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-:-KFj2Tsz4DyGYxRL_1O5:847014335)*
<!-- Sent from Reviewable.io -->

tamird commented:
Just to bring up the idea of a central (temporary) rebalancing master which was discussed in the original rebalancing RFC. Seems to me that we're gonna want to do that soon (though I'm fine with this going in as an incremental improvement first).

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/allocator.go, line 322 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj7vzFlq2JQMZZbQc4:-KFj7vzGCKstj2J2ckO3:451531353) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L322)):*
nit: you can let the default formatter do its thing: `log.Infof("ineligible for rebalance for %s", a.nextRebalance.Sub(now))`

---

*[storage/allocator.go, line 351 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8HJocMepV5zNlsp6:-KFj8HJocMepV5zNlsp7:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L351)):*
ditto

---

*[storage/balancer.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8QjzEc59FDmtLJqw:-KFj8QjzEc59FDmtLJqx:1036490910) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/balancer.go#L43)):*
one type per line

---

*[storage/balancer.go, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8S9b614un_vhclLM:-KFj8S9b614un_vhclLN:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/balancer.go#L78)):*
ditto

---

*[storage/balancer.go, line 122 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8n7DpWYBNBJ8UH8M:-KFj8n7E9bFonZQ5VBWc:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/balancer.go#L122)):*
ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-:-KFj8weRyL5GMuD66Z8j:-538472308)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
@tamird: yes, as we discussed in person, having a centralized allocators has merits that will be detailed in an upcoming RFC

---

Review status: 5 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/allocator.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFfC_ys5EFUmT-ZUVlI-r1-49:-KFjaeNGl6RLKdjZIgpI:-1828327500) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L49)):*
`gossipStoresInterval` is in the `server` package, which imports `storage`. So, I can't reference `server.gossipStoresInterval` from `storage` without a circular dependency. We could solve this by moving `gossipStoresInterval` to some third package, but I don't see an obvious choice.

Good suggestion for environment variables, which I meant to do but forgot. I've added environment variables for `gossipStoresInterval` as well as variables for min/max rebalance intervals.

---

*[storage/allocator.go, line 322 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj7vzFlq2JQMZZbQc4:-KFjvhOZFq9vnjRzlwG3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L322)):*
Done.

---

*[storage/allocator.go, line 351 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8HJocMepV5zNlsp6:-KFjvhjgrzOJUA-KDvV0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L351)):*
Done.

---

*[storage/balancer.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8QjzEc59FDmtLJqw:-KFjw3oBAB7ATqkA3YkL:-647899235) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/balancer.go#L43)):*
Doesn't the first example of the style guide have something with the same style I used?

https://github.com/cockroachdb/cockroach/blob/master/CONTRIBUTING.md#style-guide

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-:-KFjzfD-Ft-DwrbrU4vM:-1745776748)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/allocator.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFfC_ys5EFUmT-ZUVlI-r1-49:-KFj0i8tdaZyVzightKC:206213401) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L49)):*
Could you put gossipStoresInterval in gossip? Seems natural enough.

---

*[storage/allocator.go, line 351 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8HJocMepV5zNlsp6:-KFk1QeWiUq9yfmW_mJe:2020059953) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L351)):*
I think it all fits on one line now

---

*[storage/balancer.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8QjzEc59FDmtLJqw:-KFk1q_tLrCkqzgNyuBP:1117272772) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/balancer.go#L43)):*
Yep, you're right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 5 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/allocator.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFfC_ys5EFUmT-ZUVlI-r1-49:-KFk4xehORVTxAf51RmS:743789071) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L49)):*
Ah right, should've thought of that. Moved to `gossip/gossip.go`

---

*[storage/allocator.go, line 351 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFj8HJocMepV5zNlsp6:-KFk5FN0VqFUmFO3ps98:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L351)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/allocator.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFfC_ys5EFUmT-ZUVlI-r1-49:-KFkAZ4CsIhEvWbESTfn:131572809) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L49)):*
K, so this constant should be defined in terms of that now, yes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133)*
<!-- Sent from Reviewable.io -->

Commit:
storage: improve replica rebalancing

Previously, the replica allocator was choosing rebalancing targets based
on a combination of used store capacity and range counts. In practice,
for small clusters, this didn't rebalance any replicas.

This change makes the allocator spread replicas as evenly as possible
across the nodes that have sufficient available capacity.

I've also added intentional delays between rebalances, because it doesn't
make much sense to do more than one rebalance per store gossip period
(currently set to 1 minute). Any more than that risks unnecessary
replica rebalancing caused by outdated gossipped StoreDescriptors.

Resolves #5692.

cuongdo commented:
Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/allocator.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-KFfC_ys5EFUmT-ZUVlI-r1-49:-KFkDJTwOypGV7aN6H0D:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fc3e3fca43ab474ab08ef8d1a13d88d9583332c/storage/allocator.go#L49)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
@BramGruneir: I've added a test for the rebalancing backoff.

We definitely could use more tests, but I think we should link those with the simulator enhancements. I was thinking we could embed the simulator in our tests and run simple scenarios through that as part of our tests.

I'd like to get this in for the next weekly release, so is there any more feedback?

---

Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-:-KFkSak4aME6fAdlkw9d:-1785935940)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
LGTM

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6133#-:-KFkWryYHWsXleZp1IQO:676834233)*
<!-- Sent from Reviewable.io -->
